snap/quota/resources.go

```diff
@@ -0,0 +1,123 @@
+	// case the memory struct is provided. Remember that we have omitempty set on
+	// the memory value in QuotaResources.
+	Limit quantity.Size `json:"limit"`
+}
+
+// Resources is built up of multiple quota limits. Each quota limit is a pointer
+// value to indicate that their presence may be optional, and because we want to detect
+// whenever someone changes a limit to '0' explicitly.
+type Resources struct {
+	Memory *ResourceMemory `json:"memory,omitempty"`
+}
+
+func (qr *Resources) validateMemoryQuota() error {
+	// make sure the memory limit is not zero
+	if qr.Memory.Limit == 0 {
+		return fmt.Errorf("cannot create quota group with no memory limit set")
+	}
+
+	// make sure the memory limit is at least 4K, that is the minimum size
+	// to allow nesting, otherwise groups with less than 4K will trigger the
+	// oom killer to be invoked when a new group is added as a sub-group to the
+	// larger group.
+	if qr.Memory.Limit <= 4*quantity.SizeKiB {
+		return fmt.Errorf("memory limit %d is too small: size must be larger than 4KB", qr.Memory.Limit)
+	}
+	return nil
+}
+
+// Validate performs validation of the provided quota resources for a group. Its intended
+// use is before creating a group.
+func (qr *Resources) Validate() error {
+	if qr.Memory == nil {
+		return fmt.Errorf("quota group must have a memory limit set")
+	}
+
+	if qr.Memory != nil {
+		if err := qr.validateMemoryQuota(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ValidateChange performs validation of new quota limits against the current limits.
+// This is to catch issues where we want to guard against lowering limits where not supported
+// or to make sure that certain/all limits are not removed.
+func (qr *Resources) ValidateChange(newLimits Resources) error {
+
+	// Check that the memory limit is not being decreased, but we allow it to be removed
+	if newLimits.Memory != nil {
+		if newLimits.Memory.Limit == 0 {
+			return fmt.Errorf("cannot remove memory limit from quota group")
+		}
+
+		// we disallow decreasing the memory limit because it is difficult to do
+		// so correctly with the current state of our code in
+		// EnsureSnapServices, see comment in ensureSnapServicesForGroup for
+		// full details
+		if qr.Memory != nil && newLimits.Memory.Limit < qr.Memory.Limit {
+			return fmt.Errorf("cannot decrease memory limit, remove and re-create it to decrease the limit")
+		}
+	}
+
+	return nil
+}
+
+// Change updates the current quota limits with the new limits. Additional verification
+// logic exists for this operation compared to when setting initial limits. Some changes
+// of limits are not allowed.
+func (qr *Resources) Change(newLimits Resources) error {
+	if err := qr.ValidateChange(newLimits); err != nil {
+		return err
+	}
+
+	if newLimits.Memory != nil {
+		qr.Memory = newLimits.Memory
+	}
+	return nil
+}
+
+func CreateResources(memoryLimit quantity.Size) Resources {
```

Comment:
s/CreateResources/NewResources/ to match the unwritten pattern of using New* for creating things, eg. bytes.NewReader/NewBuffer, http.NewRequest, or quota.NewGroup()

Reply:
Yea I completely agree on this, I don't know why I didn't do this. I prefer the New too, and it should match the environment